content/en/blog/_posts/2022-10-27-live-and-let-live-with-kluctl-and-ssa.md

```diff
@@ -0,0 +1,174 @@
+`PATCH` request, which signals
+that Kluctl wants to perform a server-side apply operation. The API server then
+responds with an OK response (HTTP status code 200), or with a Conflict response
+(HTTP status 409).
+
+In case of a Conflict response, the body of that response includes machine readable
+details about the conflicts. Kluctl can then use these details to figure out
+which fields are in conflict and which actors (field managers) have taken
+ownership of the conflicted fields.
+
+Then, for each field, Kluctl will decide if the conflict should be ignored or
+if it should be force-applied. If any field needs to be force-applied, Kluctl
+will retry the apply operation with the ignored fields omitted and the `force`
+flag being set on the API call.
+
+In case a conflict is ignored, Kluctl will issue a warning to the user so that
+the user can properly react (or ignore it forever...).
+
+That's basically it. That is all that is required to leverage server-side apply.
+Big thanks and thumbs-up to the Kubernetes developers who made this possible!
+
+## Deciding which conflicts to ignore and which to force-apply
+
+Kluctl has a few simple rules to figure out if a conflict should be ignored
+or force-applied.
+
+It first checks the field's actor (the field manager) against a list of known
+tools that are well known to be used to perform manual modifications. These
+are for example `kubectl` and `k9s`. Any modifications performed with these tools
+are considered "temporary" and will be overwritten by Kluctl.
+
+If the field manager is not known by Kluctl, it will check if a force-apply is
+requested for that field. A force-apply can be requested in different ways:
+
+1. By passing `--force-apply` to Kluctl. This will cause ALL fields to be force-applied on conflicts.
+2. By adding the [`kluctl.io/force-apply=true`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply) annotation to the object in question. This will cause all fields of that object to be force-applied on conflicts.
+3. By adding the [`kluctl.io/force-apply-field=my.json.path`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply-field) annotation to the object in question. This cause only fields matching the JSON path to be force-applied on conflicts.
+
+## Future work in regard to conflict resolution
+
+I plan to add more ways to control the behaviour of conflict resolution. For
+example:
+
+1. It will be possible to force-apply on field level via the CLI.
+2. It will be possible to force-apply on field level but only for specific field managers.
+3. It will be possible to modify the list of known actors for which fields are always force-applied.
+4. Exclude fields from force-apply even if the field manager is in the list of well known tools.
+5. If more is needed, you can [open an issue](https://github.com/kluctl/kluctl/issues).
```

Comment:
Would you be willing @codablock to add your tips for other people who are developing tools in the same space?

I'd like this article to speak to that audience as well as to people who have selected Kluctl as their favored solution.

Reply:
Done as part of the "A few takeaways" chapter. I also removed the "Future work in regard to conflict resolution" chapter and integrated a small part of it into the previous chapter.